### PR TITLE
Auto-convert GPU arrays that support the __cuda_array_interface__ protocol

### LINKED ISSUE #15601: Auto-convert GPU arrays that support the __cuda_array_interface__ protocol
A few projects now implement the `__cuda_array_interface__` protocol, which was originally designed by the Numba developers to enable clean operation on arrays managed by other projects.  I believe that today both PyTorch and CuPy have implemented this protocol, and that Numba is the only consumer.

However, libraries like PyTorch and CuPy could also identify this protocol on input objects and use it within their functions.  Probably the first use case would be to allow users to easily convert between different GPU array libraries.  For example, ideally the following would work.

```python
import cupy
x = cupy.random.random((1000, 1000))

import torch
t = torch.tensor(x)
```

Ideally the check within `torch.tensor` would be something like the following:

```python
if hasattr(x, '__cuda_array_interface__'):
    ...
```

rather than the following:

```python
if isinstance(x, cupy.ndarray):
    ...
```

which would be useful for support of future GPU libraries.

xref https://github.com/pytorch/pytorch/issues/11914

### PR DESCRIPTION
**Pull request by madsbk**
This PR implements auto-conversion of GPU arrays that support the `__cuda_array_interface__` protocol (fixes #15601).

If an object exposes the `__cuda_array_interface__` attribute, `touch.as_tensor()` and `touch.tensor()` will use the exposed device memory.

#### Zero-copy 
When using `touch.as_tensor(...,device=D)` where `D` is the same device as the one used in `__cuda_array_interface__`.

#### Implicit copy 
When using `touch.as_tensor(...,device=D)` where `D` is the CPU or another non-CUDA device.

#### Explicit copy 
When using `torch.tensor()`.

#### Exception
When using `touch.as_tensor(...,device=D)` where `D` is a CUDA device not used in `__cuda_array_interface__`.

#### Lifetime 
`torch.as_tensor(obj)` tensor grabs a reference to `obj` so that the lifetime of `obj` exceeds the tensor

### CODE REVIEW

File: test/test_numba_integration.py
Comment (mrocklin) [inline test/test_numba_integration.py]: If easy, I suggest that we also test what happens when we delete the numba device array.  Is the memory in the Torch tensor still valid?  (I think that the answeer is yes, but this was a concern originally, so it would be good to verify that it works)

Comment (madsbk) [reply test/test_numba_integration.py]: Good idea, added in 4742f91

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp]: Is there any reason to only enable this when we compile with `USE_NUMPY`? It seems to me that it should work even without Numpy.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: I'm going to ask you to port the Python bindings here to use pybind11. The reason for this is that you are leaking memory in error conditions (I'll point this out shortly), and if you use actual RAII C++ objects, these problems will go away. It's much more scalable than having me manually run a "did you free the refcounted object on all logic paths" analysis on your code.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Leaks `cuda_dict`

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: This "looks" as if it frees `cuda_dict` on error conditions, but actually because you are only catching `ValueError`, all of the `TypeError` exit points leak memory. To make matters worse, you never free `cuda_dict` on the success case; even more leaky memory!

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: You have an owning reference and must free it on all exit paths (the easiest way to make sure this happens is to use one of pybind11's RAII owning reference types!)

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: This isn't just a TypeError; if this ever fails, that's an internal error (the precondition for this function is that `__cuda_array_interface__` exists, right?) So a `TORCH_INTERNAL_ASSERT` is an appropriate thing to use to check the nullptr here.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: I don't think there's any reason to use a `npy_intp` here, a `void*` will express this perfectly well, I think.

Comment (ezyang) [reply torch/csrc/utils/tensor_numpy.cpp]: (Or maybe a `intptr_t`, if you actually need a signed integral type. I haven't read the rest of the code yet, so I don't know exactly what you need.)

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: This use of the Numpy API doesn't seem necessary. You've got a sequence, just read the numbers out the old fashioned way!

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: That would save you from having to do another conversion to actually get into ATen sizes.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Because `assert` cause a terminate() on failure, we don't use them in PyTorch codebase.  Use `TORCH_INTERNAL_ASSERT` to specify that something should be true (and a user should report a bug if it's not true.)

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: OK, this seems like a legit use of the Numpy API.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [reply torch/csrc/utils/tensor_numpy.cpp]: Update: Since we have to use Numpy anyway for descriptors, this code is much less bad. Downgrade this comment to a "nit".

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: I'm not sure why you need a `PyErr_Clear()` here

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: nit: This is another use of the Numpy API which doesn't seem necessary to implement the functionality at hand

**Review by ezyang — CHANGES_REQUESTED** [review]
Logically, the code looks great! However, I would like the memory leaks to be fixed before merge. If you insist, just manually inserting the necessary decrefs is acceptable, however, I think using an RAII class from pybind11 will be much safer.

**Comment by madsbk** [comment]
Thanks, I will use pybind11 to fix the memory leaks and also address the other issues.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (madsbk) [reply torch/csrc/utils/tensor_numpy.cpp]: Fixed in 5b0c83c

File: torch/csrc/utils/tensor_numpy.cpp
Comment (madsbk) [reply torch/csrc/utils/tensor_numpy.cpp]: Fixed in ff77732

File: torch/csrc/utils/tensor_numpy.cpp
Comment (madsbk) [reply torch/csrc/utils/tensor_numpy.cpp]: Fixed in 2cf1117

File: torch/csrc/utils/tensor_numpy.cpp
Comment (madsbk) [reply torch/csrc/utils/tensor_numpy.cpp]: Fixed in 2cf1117

File: torch/csrc/utils/tensor_numpy.cpp
Comment (madsbk) [reply torch/csrc/utils/tensor_numpy.cpp]: Fixed in 5b0c83c

File: torch/csrc/utils/tensor_numpy.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: I don't think you intended to have this line?

**Review by ezyang — APPROVED** [review]
Will merge when tests pass.

**Review by ezyang — APPROVED** [review]

**Comment by mrocklin** [comment]
FYI @seibert, thought you'd like to know that PyTorch now supports `__cuda_array_interface__` in both directions

**Pull request closed without merging**
